### ISSUE TITLE
Prevent tables smaller than one page from spanning two pages in PDF output

### ISSUE THREAD

**Issue by reubano** [jgm/pandoc#782]
Moving from ctable to longtable causes page breaks in short tables when they are at the bottom of the page. It would be better to move the table to the next page instead. Or at least give the user an option.

https://groups.google.com/d/msg/pandoc-discuss/pmiw78NvZl4/RMnCVKJr-k4J

**Comment by ickc** [comment]
I face a similar problem too that a very small table is spread in 2 pages, and using `\floatbarrier` didn't fix it since it is not because it is floating. Do any one know a quick fix (except for embedding raw LaTeX table)?

**Comment by FlxVctr** [comment]
same problem

**Comment by ickc** [comment]
> same problem

Do you know you just notified every one in this thread? Think before you post next time. “Same problem” isn’t going to help anyone but to spam everyone in a thread.

**Comment by davidskalinder** [comment]
Kinda surprised this one hasn't had more interest over thirteen years, since splitting a small table over two pages seems like a showstopper for respectable table formatting.  Maybe everyone else knows of a good solution to this that I didn't find?

Anyway in case it's useful to anyone, the following is a Lua filter that works around the problem by simply wrapping longtables (or any block) with the class `.wrapminipage` into a minipage to keep it from breaking across pages.

Note that if this is used through Quarto, the classes seem to get juggled around while Quarto processes the AST, so I've found that it needs the [at: post-render](https://quarto.org/docs/prerelease/1.4/lua_changes.html#more-precise-targeting-of-ast-processing-phases) option set in order to find the class reliably.

```
local function wrap_minipage (el)
  if (
    FORMAT:match("latex")
    and el.classes ~= nil
    and el.classes:includes("wrapminipage")
  ) then
    eldoc = pandoc.Pandoc(el)
    local wrappedtex = pandoc.write(eldoc, "latex"):gsub(
      ".*", "\\begin{minipage}{\\textwidth}%0\\end{minipage}")
    local new_el = pandoc.RawBlock("tex", wrappedtex)
    return new_el
  else
    return el
  end
end

function Pandoc (doc)
  doc = doc:walk { Block = wrap_minipage }
  return doc
end

```

Without this filter, rendering the doc below to a PDF splits the table over the page, but it stays together with the filter.

<details>
<summary>(Click to expand longish doc source)</summary>

```
---
title: "Untitled"
---

a

b

c

d

e

f

g

h

i

j

k

l

m

n

o

p

q

r

s

t

::: {.wrapminipage}

col1 | col2
-- | --
11111111111111111111111 | 111
22222222222222222222222 | 222
33333333333333333333333 | 333
44444444444444444444444 | 444

A caption for this table.
:::

```

</details>

EDIT: An example of usage in a Quarto doc:

<details>
<summary>(Click to expand longish doc source)</summary>

```
---
title: "Untitled"
format:
  pdf: default
filters:
  - at: post-render
    path: filters/wrapminipage.lua
---

{{< lipsum 3 >}}

::: {#tbl-whatever .wrapminipage}

col1 | col2
-- | --
11111111111111111111111 | 111
22222222222222222222222 | 222
33333333333333333333333 | 333
44444444444444444444444 | 444

A caption for this table.
:::


{{< lipsum 1 >}}

```

</details>

**Comment by jgm** [comment]
What we ideally want is:

- to use a *floating* table environment like regular `table` when the table will fit on one page (the non-floating minipage approach is likely to lead to big whitespace gaps at the bottom of a page)
- use the (non-floating) `longtable` otherwise

But I don't know how we can tell whether it will fit one page.

**Comment by davidskalinder** [comment]
> the non-floating minipage approach is likely to lead to big whitespace gaps at the bottom of a page

Yes, good point. I think the filter above is good enough for me for the moment, but it could be adapted without too much trouble to `gsub` out all the `longtable` stuff and put in `table`/`tabular` stuff instead.

But of course that doesn't help with

> how we can tell whether it will fit one page.

The first thing that occurs to me is to look at line counts, but it wouldn't surprise me if that would quickly run into complications that you understand a lot better than I do. :)

**Comment by davidskalinder** [comment]
> `gsub` out all the `longtable` stuff

Which for reference (looking at the [writer source](https://github.com/jgm/pandoc/blob/d39ca219e7b8de7fee88dd46a6335c17419fecc1/src/Text/Pandoc/Writers/LaTeX/Table.hs)) looks like it includes:

```
\\LTcaptype{none}
\\begin{longtable}[]
\\caption
\\endfirsthead
\\endhead
\\endlastfoot
\\end{longtable}
```

The hard(er) part is to move the caption outside of the `tabular` environment (and possibly decide whether to put it before or after)?

All this would just be for improving the workaround of course, not for the real solution.

**Comment by jgm** [comment]
One idea would be to check for the class `.float` on a table and generate a table > tabular floating table instead of a longtable if it is present.  This would allow the user to control whether a floating table or a longtable is used.
